### PR TITLE
Rad 5452: Back port of MANGO-1051. Enable use of OIDC/OAuth JWT bearer tokens

### PR DESCRIPTION
**Pull request by ManuelCalvo3Pillar**
RAD-5452: Back port of MANGO-1051. Enable use of OIDC/OAuth JWT bearer tokens.

Changes:
* Leaving BearerAuthenticationFilter class as requested.
* Leaving BearerAuthenticationToken class as requested.
* Converting Java records to read only VO  in JwtPermissionHolder.java and JwtTokenAuthenticationProvider.java
* Removed the usage of Clock in JwtSignerVerifier, because we are not back porting unit tests, and this was only needed for that.
* This is not supported in Java 11 `"string%s".formatted(scope)` changing to `String.format("string%s", scope)`  in JwtTokenAuthenticationProvider.java
* This is not supported in Java 11 
`if (!(authentication instanceof BearerTokenAuthenticationToken bearerToken)) return null;`
changing to:
```
if (!(authentication instanceof BearerTokenAuthenticationToken)) return null;
BearerTokenAuthenticationToken bearerToken = (BearerTokenAuthenticationToken) authentication;
```
in MangoTokenAuthenticationProvider.java and JwtTokenAuthenticationProvider.java
* Removing MangoTokenAuthenticationProviderTest as unit tests won't be back ported.

### CODE REVIEW

File: Core/src/com/serotonin/timer/SimulationTimer.java [hunk 1]
```diff
@@ -70,6 +72,23 @@ public void next() {
...
+    }
+
+    /**
+     * @param time the new instant to set the clock to
+     * @return the current instant (same as {@link #instant() would return}
+     */
+    public Instant fastForwardTo(Instant time) {
+        fastForwardTo(time.toEpochMilli());
+        return time;
+    }
```
Comment (jazdw) [inline Core/src/com/serotonin/timer/SimulationTimer.java:90]: New API in a patch release, remove this, it's not required.

File: Core/src/com/serotonin/m2m2/web/mvc/spring/security/authentication/BearerAuthenticationToken.java
Comment (jazdw) [inline Core/src/com/serotonin/m2m2/web/mvc/spring/security/authentication/BearerAuthenticationToken.java:1]: Deleted API in a patch release, shouldn't affect anything, but might be safer to leave this class?

File: Core/src/com/serotonin/m2m2/web/mvc/spring/security/BearerAuthenticationFilter.java
Comment (jazdw) [inline Core/src/com/serotonin/m2m2/web/mvc/spring/security/BearerAuthenticationFilter.java:1]: Deleted API in a patch release, shouldn't affect anything, but might be safer to leave this class?

File: Core/src/com/infiniteautomation/mango/spring/components/TokenAuthenticationService.java [hunk 2]
```diff
@@ -51,7 +52,12 @@ public final class TokenAuthenticationService extends JwtSignerVerifier<User> {
     public TokenAuthenticationService(
             PermissionService permissionService,
             UsersService usersService,
-            ApplicationContext context, RunAs runAs, UserDao userDao) {
+            ApplicationContext context,
+            RunAs runAs,
+            UserDao userDao,
+            Clock clock) {
```
Comment (jazdw) [inline Core/src/com/infiniteautomation/mango/spring/components/TokenAuthenticationService.java:58]: Modified API in a patch release, might be safer to not backport the tests (this change was for testing only).

File: Core/src/com/infiniteautomation/mango/spring/components/PasswordResetService.java [hunk 3]
```diff
@@ -65,7 +68,13 @@ public final class PasswordResetService extends JwtSignerVerifier<User> {
     public PasswordResetService(
             PermissionService permissionService,
             UsersService usersService,
-            PublicUrlService publicUrlService, RunAs runAs, PageResolver pageResolver, SystemSettingsDao systemSettingsDao) {
+            PublicUrlService publicUrlService,
+            RunAs runAs,
+            PageResolver pageResolver,
+            SystemSettingsDao systemSettingsDao,
+            Clock clock) {
```
Comment (jazdw) [inline Core/src/com/infiniteautomation/mango/spring/components/PasswordResetService.java:75]: Modified API in a patch release, might be safer to not backport the tests (this change was for testing only).

File: Core/src/com/infiniteautomation/mango/spring/components/EmailAddressVerificationService.java [hunk 4]
```diff
@@ -75,7 +76,12 @@ public EmailAddressVerificationService(
             UsersService usersService,
             PublicUrlService publicUrlService,
             SystemSettingsDao systemSettings,
-            PermissionService permissionService, RunAs runAs, PageResolver pageResolver) {
+            PermissionService permissionService,
+            RunAs runAs,
+            PageResolver pageResolver,
+            Clock clock) {
```
Comment (jazdw) [inline Core/src/com/infiniteautomation/mango/spring/components/EmailAddressVerificationService.java:82]: Modified API in a patch release, might be safer to not backport the tests (this change was for testing only).

File: Core/src/com/infiniteautomation/mango/jwt/JwtSignerVerifier.java [hunk 5]
```diff
@@ -47,36 +48,39 @@ public abstract class JwtSignerVerifier<T> {
...
 
+    private final Clock clock;
     private KeyPair keyPair;
     private JwtParser parser;
 
     protected final Logger log;
 
-    protected JwtSignerVerifier() {
+    protected JwtSignerVerifier(Clock clock) {
+        this.clock = clock;
```
Comment (jazdw) [inline Core/src/com/infiniteautomation/mango/jwt/JwtSignerVerifier.java:58]: Modified API in a patch release, might be safer to not backport the tests (this change was for testing only).

File: Core/src-test/com/serotonin/m2m2/web/mvc/spring/security/authentication/MangoTokenAuthenticationProviderTest.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+public class MangoTokenAuthenticationProviderTest extends MangoTestBase {
+
+  AuthenticationManager authenticationManager;
+  TokenAuthenticationService tokenAuthService;
+  UsersService usersService;
+  SimulationTimer clock;
+
+  @Override
+  protected MockMangoLifecycle getLifecycle() {
+    return super.getLifecycle().setInitializeWebServer(true);
```
Comment (jazdw) [inline Core/src-test/com/serotonin/m2m2/web/mvc/spring/security/authentication/MangoTokenAuthenticationProviderTest.java:48]: Does this method (`setInitializeWebServer`) actually exist in 4.5.x?

**Review by jazdw — CHANGES_REQUESTED** [review]
Please write a summary of what you had to modify to backport this. e.g. convert record to class etc.

@terrypacker Can you review, there are breaking API changes which is not something we tried very hard to avoid in patch releases. Some are probably inevitable here though.

**Review by benji811124 — APPROVED** [review]
I have reviewed the changes and looks good to me

File: Core/src/com/serotonin/timer/SimulationTimer.java (hunk 1 above)
Comment (ManuelCalvo3Pillar) [reply Core/src/com/serotonin/timer/SimulationTimer.java:90]: Ok, removing it.

File: Core/src/com/serotonin/m2m2/web/mvc/spring/security/authentication/BearerAuthenticationToken.java
Comment (ManuelCalvo3Pillar) [reply Core/src/com/serotonin/m2m2/web/mvc/spring/security/authentication/BearerAuthenticationToken.java:1]: Ok, adding it back. Yes, I think so.

File: Core/src/com/serotonin/m2m2/web/mvc/spring/security/BearerAuthenticationFilter.java
Comment (ManuelCalvo3Pillar) [reply Core/src/com/serotonin/m2m2/web/mvc/spring/security/BearerAuthenticationFilter.java:1]: @jazdw if I leave it there, wouldn't be it executed because it is a filter?

File: Core/src/com/infiniteautomation/mango/spring/components/TokenAuthenticationService.java (hunk 2 above)
Comment (ManuelCalvo3Pillar) [reply Core/src/com/infiniteautomation/mango/spring/components/TokenAuthenticationService.java:58]: Ok, undoing it.

File: Core/src/com/infiniteautomation/mango/spring/components/PasswordResetService.java (hunk 3 above)
Comment (ManuelCalvo3Pillar) [reply Core/src/com/infiniteautomation/mango/spring/components/PasswordResetService.java:75]: Ok, undoing it.

File: Core/src/com/infiniteautomation/mango/spring/components/EmailAddressVerificationService.java (hunk 4 above)
Comment (ManuelCalvo3Pillar) [reply Core/src/com/infiniteautomation/mango/spring/components/EmailAddressVerificationService.java:82]: Ok, undoing it.

File: Core/src/com/infiniteautomation/mango/jwt/JwtSignerVerifier.java (hunk 5 above)
Comment (ManuelCalvo3Pillar) [reply Core/src/com/infiniteautomation/mango/jwt/JwtSignerVerifier.java:58]: Ok, undoing it.

File: Core/src/com/serotonin/m2m2/web/mvc/spring/security/BearerAuthenticationFilter.java
Comment (jazdw) [reply Core/src/com/serotonin/m2m2/web/mvc/spring/security/BearerAuthenticationFilter.java:1]: It's not annotated as a `@Component` so it shouldn't be picked up by Spring automatically. It used to be manually added in `MangoSecurityConfiguration`

Comment (ManuelCalvo3Pillar) [reply Core/src/com/serotonin/m2m2/web/mvc/spring/security/BearerAuthenticationFilter.java:1]: Got it

File: Core/src-test/com/serotonin/m2m2/web/mvc/spring/security/authentication/MangoTokenAuthenticationProviderTest.java (hunk 6 above)
Comment (ManuelCalvo3Pillar) [reply Core/src-test/com/serotonin/m2m2/web/mvc/spring/security/authentication/MangoTokenAuthenticationProviderTest.java:48]: @jazdw yes it does, I was able to run successfully the test before these changes.

File: Core/src/com/infiniteautomation/mango/jwt/JwtSignerVerifier.java
Comment (jazdw) [inline Core/src/com/infiniteautomation/mango/jwt/JwtSignerVerifier.java:1]: Revert these changes

File: Core/src/com/infiniteautomation/mango/spring/components/TokenAuthenticationService.java
Comment (jazdw) [inline Core/src/com/infiniteautomation/mango/spring/components/TokenAuthenticationService.java:1]: Revert these changes

File: Core/src/com/infiniteautomation/mango/spring/components/PasswordResetService.java
Comment (jazdw) [inline Core/src/com/infiniteautomation/mango/spring/components/PasswordResetService.java:1]: Revert these changes

**Review by jazdw — CHANGES_REQUESTED** [review]

File: Core/src/com/infiniteautomation/mango/spring/components/TokenAuthenticationService.java
Comment (ManuelCalvo3Pillar) [reply Core/src/com/infiniteautomation/mango/spring/components/TokenAuthenticationService.java:1]: @jazdw can you be more specific? I reverted as requested, do you want to add back the Clock?

File: Core/src/com/infiniteautomation/mango/jwt/JwtSignerVerifier.java
Comment (ManuelCalvo3Pillar) [reply Core/src/com/infiniteautomation/mango/jwt/JwtSignerVerifier.java:1]: @jazdw can you be more specific? I reverted as requested, do you want to add back the Clock?

File: Core/src/com/infiniteautomation/mango/spring/components/PasswordResetService.java
Comment (ManuelCalvo3Pillar) [reply Core/src/com/infiniteautomation/mango/spring/components/PasswordResetService.java:1]: @jazdw can you be more specific? I reverted as requested, do you want to add back the Clock?

File: Core/src/com/infiniteautomation/mango/jwt/JwtSignerVerifier.java
Comment (ManuelCalvo3Pillar) [reply Core/src/com/infiniteautomation/mango/jwt/JwtSignerVerifier.java:1]: Ok done.

File: Core/src/com/infiniteautomation/mango/spring/components/TokenAuthenticationService.java
Comment (ManuelCalvo3Pillar) [reply Core/src/com/infiniteautomation/mango/spring/components/TokenAuthenticationService.java:1]: ok, done

File: Core/src/com/infiniteautomation/mango/spring/components/PasswordResetService.java
Comment (ManuelCalvo3Pillar) [reply Core/src/com/infiniteautomation/mango/spring/components/PasswordResetService.java:1]: The IDE automatically re-arranges the imports, I think as per Mango wiki config. Is it too much of a problem that?

File: Core/src/com/infiniteautomation/mango/spring/components/EmailAddressVerificationService.java
Comment (jazdw) [inline Core/src/com/infiniteautomation/mango/spring/components/EmailAddressVerificationService.java:1]: Revert all changes to this file

File: Core/src/com/infiniteautomation/mango/spring/components/PasswordResetService.java
Comment (jazdw) [inline Core/src/com/infiniteautomation/mango/spring/components/PasswordResetService.java:1]: Revert all changes to this file

File: Core/src/com/infiniteautomation/mango/spring/components/TokenAuthenticationService.java
Comment (jazdw) [inline Core/src/com/infiniteautomation/mango/spring/components/TokenAuthenticationService.java:1]: Revert all changes to this file

File: Core/src/com/serotonin/m2m2/web/mvc/spring/security/authentication/MangoTokenAuthenticationProvider.java [hunk 7]
```diff
@@ -77,14 +76,14 @@ public Authentication authenticate(Authentication authentication) throws Authent
         userDetailsChecker.check(user);
 
         if (log.isDebugEnabled()) {
-            log.debug("Successfully authenticated user using JWT token, header: " + jws.getHeader() + ", body: " + jws.getBody());
+            log.debug(String.format("Successfully authenticated user using JWT token, header: %s, body: %s", jws.getHeader(), jws.getBody()));
```
Comment (jazdw) [inline Core/src/com/serotonin/m2m2/web/mvc/spring/security/authentication/MangoTokenAuthenticationProvider.java:79]: I would rather this change was not made, I know the IDE is probably warning you about string concatenation on this line but there are 2 reasons to not touch this:
- If we ever change this in 5.x and need to backport again there will be a conflict
- Its not the correct fix, we should use `log.debug("Successfully authenticated user using JWT token, header: {}, body: {}", jws.getHeader(), jws.getBody());` (this is actually the fix that IntelliJ suggests, see screenshot)

image

**Review by jazdw — CHANGES_REQUESTED** [review]
Rule of thumb: Change as little as possible when doing a backport, keep changes as 1:1 as possible with the original commit.

File: Core/src/com/infiniteautomation/mango/spring/components/EmailAddressVerificationService.java
Comment (ManuelCalvo3Pillar) [reply Core/src/com/infiniteautomation/mango/spring/components/EmailAddressVerificationService.java:1]: ok, done

File: Core/src/com/infiniteautomation/mango/spring/components/PasswordResetService.java
Comment (ManuelCalvo3Pillar) [reply Core/src/com/infiniteautomation/mango/spring/components/PasswordResetService.java:1]: ok done

File: Core/src/com/infiniteautomation/mango/spring/components/TokenAuthenticationService.java
Comment (ManuelCalvo3Pillar) [reply Core/src/com/infiniteautomation/mango/spring/components/TokenAuthenticationService.java:1]: ok, done

File: Core/src/com/serotonin/m2m2/web/mvc/spring/security/authentication/MangoTokenAuthenticationProvider.java (hunk 7 above)
Comment (ManuelCalvo3Pillar) [reply Core/src/com/serotonin/m2m2/web/mvc/spring/security/authentication/MangoTokenAuthenticationProvider.java:79]: ok, done

File: Core/src/com/infiniteautomation/mango/spring/components/EmailAddressVerificationService.java
Comment (jazdw) [reply Core/src/com/infiniteautomation/mango/spring/components/EmailAddressVerificationService.java:1]: Not done

File: Core/src/com/infiniteautomation/mango/spring/components/PasswordResetService.java
Comment (jazdw) [reply Core/src/com/infiniteautomation/mango/spring/components/PasswordResetService.java:1]: Not done

File: Core/src/com/infiniteautomation/mango/spring/components/TokenAuthenticationService.java
Comment (jazdw) [reply Core/src/com/infiniteautomation/mango/spring/components/TokenAuthenticationService.java:1]: Not done

File: Core/src/com/serotonin/m2m2/web/mvc/spring/security/authentication/MangoTokenAuthenticationProvider.java (hunk 7 above)
Comment (jazdw) [reply Core/src/com/serotonin/m2m2/web/mvc/spring/security/authentication/MangoTokenAuthenticationProvider.java:79]: Not done

**Review by jazdw — CHANGES_REQUESTED** [review]

**Review by jazdw — APPROVED** [review]
